### PR TITLE
Feature/issue 871 (setCellHeaderProps) and Feature/issue 714 (setTableProps)

### PR DESCRIPTION
**Pull request by patorjk**
https://github.com/gregnb/mui-datatables/issues/871 and https://github.com/gregnb/mui-datatables/issues/714

Added two new options:

* setTableProps (options)
* setCellHeaderProps (columns)

They're modeled after the existing setCellProps and setRowProps. The dense table is really only available in Material v4, so the example only removes some padding in Material v3.

Quick view: https://codesandbox.io/s/github/patorjk/mui-datatables/tree/set-props

### CODE REVIEW

**Comment by patorjk** [comment]
I'm glad this is on your radar. Since discovering setCellProps and setRowsProps I've found them incredibly useful. They provide a lot of power and a nice reprieve from render functions. Some examples:

* Using setCellProps I'm able to eliminate 1,700 divs from one of my tables (previously I was wrapping every cell in a div with a style of "white-space:pre").
* Using setCellProps and setCellHeaderProps I'm able to have more concise code.
* setCellProps and setCellHeaderProps make the issue discussed here: https://github.com/gregnb/mui-datatables/issues/681 - completely trivial. 

Render functions technically allow for the same thing, but I find them dreadful and they're way more complex than would seem. For example, to override a cell header, you really need to add in most of the functionality of [TableHeadCell.js](https://github.com/gregnb/mui-datatables/blob/master/src/components/TableHeadCell.js), and there's a lot going on in there. I don't think it's reasonable to assume that most devs will take the care to reproduce what's in there, leading to bugs in their code.

Anyway, some time has passed since I put this in. setTableProps may be better off as "tableProps", as there's only 1 table. Additionally, if you decided you like this type of API, you may want to reconsider the searchProps idea of https://github.com/gregnb/mui-datatables/pull/808 (though that implementation is buggy - it's very close though). searchProps also makes a lot of customization trivial, and itsolves a lot of issues people brought up in https://github.com/gregnb/mui-datatables/issues/857#issuecomment-530179241.

File: src/MUIDataTable.js [hunk 1]
```diff
@@ -1229,6 +1233,10 @@ class MUIDataTable extends React.Component {
         break;
     }
 
+    let tableProps = this.options.setTableProps ? this.options.setTableProps() : {};
```
Comment (gabrielliwerant) [inline src/MUIDataTable.js]: Maybe we should be more defensive here. In case the function doesn't return an object, it will error out below. What do you think?

File: src/components/TableBodyCell.js [hunk 2]
```diff
@@ -12,18 +12,26 @@ const defaultBodyCellStyles = theme => ({
     [theme.breakpoints.down('sm')]: {
       display: 'inline-block',
       fontSize: '16px',
-      height: '24px',
```
Comment (gabrielliwerant) [inline src/components/TableBodyCell.js]: Could we hold off on the stylistic changes here or put in another PR? I believe there were some modifications here since you opened this that might have solved some of the style issues, but also, if we keep them separate, then we can discuss them separately and then they won't hold up the other work here.

File: src/components/TableHeadCell.js [hunk 3]
```diff
@@ -89,7 +89,8 @@ class TableHeadCell extends React.Component {
 
   render() {
     const { isSortTooltipOpen, isHintTooltipOpen } = this.state;
-    const { children, classes, options, sortDirection, sort, hint, print } = this.props;
+    const { children, classes, options, sortDirection, sort, hint, print, cellHeaderProps } = this.props;
+    const { className, ...otherProps } = cellHeaderProps;
```
Comment (gabrielliwerant) [inline src/components/TableHeadCell.js]: Similar to the other comment about checking for the `className` property, as the dev will get an unhelpful error if they return the wrong type.

**Comment by gabrielliwerant** [comment]
Understood about props vs render, I agree that custom render functions are a bad option when you want to replicate the functionality almost entirely, but with some slight changes. I think it would be a nice option to allow various parts of the table to be exported separately for just this kind of usage, but that's probably a ways away. I think render functions work best when the functionality being added is very custom and much different from exists, and probably has a bunch of nested components and such. But use cases like those you mention are better served via adding props, so it's nice to have both.

**Comment by gabrielliwerant** [comment]
Hm, I don't see a big difference between `setTableProps`  and `tableProps` for the API. I think it's better to be consistent with the other options for now. Maybe there's an even better organizing principle that can be used for API naming, like how callbacks all begin with `on`. I've also been toying with the idea of adding a prefix for boolean options because as of now they are difficult to distinguish from many of the non-boolean options. So this is something I do think about.

Agreed about `searchProps` having valid use cases outside of the `customSearchRender`. I'll ping the PR creator in the next day or so and see if he wants to update it. If not, It'll be up for grabs.

File: src/MUIDataTable.js (hunk 1 above)
Comment (patorjk) [reply src/MUIDataTable.js]: I don't know, that's not how the existing setCellProps and setRowProps functions work. Trying out other return values in the console, it doesn't look like anything breaks with the spread operator, they just wouldn't be setting any fields if they had the function return a non-object (with the exception of a "string" type, where they'd set numerical fields on the component).

File: src/components/TableBodyCell.js (hunk 2 above)
Comment (patorjk) [reply src/components/TableBodyCell.js]: They're tied to the functionality a setTableProps method would add (ie, "dense table"). Without these changes, applying a dense table breaks stacked mode. There's actually bugs in the existing code, if you view the Customize Styling example you'll see one of them:

https://codesandbox.io/s/github/gregnb/mui-datatables/

The CSS solution I wound up with is simple. Initially I tried to be fancy, but it wasn't needed and didn't cover certain edge cases (basically a height parameter is needed for causes when cell values are empty or just a space).

File: src/components/TableHeadCell.js (hunk 3 above)
Comment (patorjk) [reply src/components/TableHeadCell.js]: I'm not sure I follow. Check that it's set to a string?

**Comment by patorjk** [comment]
I can circle back a little later for tests.

File: src/MUIDataTable.js (hunk 1 above)
Comment (gabrielliwerant) [reply src/MUIDataTable.js]: The others don't work this way because they aren't checking an object for a specific property. The new functions you added are checking to see if `classNames` appears on the object, and this will error for non-objects, where the others won't, with the exception of strings.

File: src/components/TableBodyCell.js (hunk 2 above)
Comment (gabrielliwerant) [reply src/components/TableBodyCell.js]: Right, so the problem is just that the PR is a two-for-one deal. In the future, I'd prefer to have them kept separate as this makes reviewing easier, but it's also much easier to trace back any future issues or regressions to a specific commit (since we squash merges) and enables us to rollback specific sets of functionality without losing others that work well, if need be.

**Comment by gabrielliwerant** [comment]
@patorjk If you want to take on `setSearchProps`, go for it. I'm going to assume the other dev that submitted the old PR isn't checking up on it, since it's gotten old and he hasn't responded to new messages.

**Comment by patorjk** [comment]
Sure, I'll put something together, I'll loop back to it later this week or next.
